### PR TITLE
Update universal intrinsics of RVV back-end.

### PR DESCRIPTION
**Pull request by hanliutong**
The goal of this patch is to optimize the existing Universal Intrinsics implementation of RVV.
The initial version of RVV UI is #18228.

The main difference between this patch and the initial version is the design of the data structure: Using a pointer to the RVV Intrinsic Type instead of an array to hold the value. In this way, no more Intrinsic functions such as load & store needed in the struct, which provides opportunities for compilation optimization.

[This](https://godbolt.org/z/4MToxzn7h) is a comparison on assembly level, taking `v_fma` as an example, the implementation of this patch reduces the number of instructions by about 50%.

This patch is tested on QEMU with the core testcases and the dnn testcases, and the results are the same as master.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/include/opencv2/core/hal/intrin_rvv.hpp [hunk 1]
```diff
@@ -531,7 +532,358 @@ struct v_float64x2
...
+    operator vuint8m1_t() const
+    {
+        return *pval;
+    }
+    uchar get0() const
+    {
+        return vmv_x(*pval);
+    }
+    inline v_uint8x16& operator=(const v_uint8x16& vec) {
+        *pval = *(vec.pval);
```
Comment (asmorkalov) [inline modules/core/include/opencv2/core/hal/intrin_rvv.hpp:561]: It's potential memory leak. pval is already initialized with malloc, see code below.

File: modules/core/include/opencv2/core/hal/intrin_rvv.hpp [hunk 2]
```diff
@@ -531,7 +532,358 @@ struct v_float64x2
...
+        return *this;
+    }
+    inline v_uint8x16(const v_uint8x16& vec) {
+        *pval = *(vec.pval);
+    }
+    vuint8m1_t* pval = (vuint8m1_t*)malloc(nlanes*8);
```
Comment (asmorkalov) [inline modules/core/include/opencv2/core/hal/intrin_rvv.hpp:567]: It's bad idea. The vector registers are usually local variables in function or several functions in call chain. Each vector variable creation cases expensive call and potentially heap fragmentation. I propose to use vuint8m1_t[nlanes*8] for the field as soon as it's always allocated/

File: modules/core/include/opencv2/core/hal/intrin_rvv.hpp [hunk 3]
```diff
@@ -531,7 +532,358 @@ struct v_float64x2
...
+        return vmv_x(*pval);
+    }
+    inline v_int8x16& operator=(const v_int8x16& vec) {
+        *pval = *(vec.pval);
+        return *this;
+    }
+    inline v_int8x16(const v_int8x16& vec) {
+        *pval = *(vec.pval);
+    }
+    vint8m1_t* pval = (vint8m1_t*)malloc(nlanes*8);
```
Comment (asmorkalov) [inline modules/core/include/opencv2/core/hal/intrin_rvv.hpp:602]: The same here.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by hanliutong** [comment]
@asmorkalov, thank you for your review and I am going to explain the design of this patch in detail about your review comment.

I agree that calling malloc for each vector variable is expensive and may fragment the heap. 

The reason I did this is to get a vector variable `*pval` in struct in order to remove unnecessary intrinsic functions for loading array and storing vector.  

Normally, for a vector variable, we would try to declare it with intrinsic type directly. But we can not declare a RVV intrinsic vector type like `vuint8m1_t val` in the struct because it is sizeless and the compiler doesn't know how many bits did it needed on the stack.  However, in OpenCV UI, we know exactly how much bit is needed, 128. 

So I tried an scalar array and a RVV type pointer like below (but not a RVV type array like `vuint8m1_t[nlanes*8]`, we may NOT need an array containing vector) , in this way, `*pval` is the vector variable that we want:
```cpp
float val[nlanes];
vfloat32m1_t* pval = (vfloat32m1_t*)val;
```
it's work but generate unecessay store instructions. And I find that use the malloc way like below can eliminate unecessay instructions, same as the array way, `*pval` is the vector variable that we want:
```cpp
vfloat32m1_t* pval = (vfloat32m1_t*)malloc(nlanes*32);
~v_float32x4() {free(pval);}
```
it also works as well as more concise assembly. And most importantly, although we call malloc in the source code, no malloc is actually called in assembly with the help of compiler optimization because (I guess) the parameters of the malloc function is a constant (128). But in more complex cases, the compiler may not be able to optimize.

https://godbolt.org/z/8sa6W8vf3 shows the difference between using a scalar array and using malloc.

Therefore, there is a tradeoff between more unnecessary store instructions (by using array) and possible malloc calls, what's your opinion?

And I would like to try to explain why I think this does not cause a memory leak:

Firstly, if there is no malloc called in assembly, then no memory in heap is created, also no way to leak.
Secondly, assuming there is a malloc called. In this case, we always operate as `*pval` and do not operate `pval` directly, which means we may change the value pointed by the pointer, but does not change the pointer itself. In other words, the pointer `pval` always points to the memory from malloc, and will be free in the destructor. 

So there does not seem to be a memory leak, please correct me if I am wrong.

**Comment by asmorkalov** [comment]
Even if compiler is smart enough and does not generate real malloc calls in some cases, it's bad idea to drop pointer values and replace them with the new one. It bad logic and it stops working. Also, it's unsafe to call free for the cases like this:
```
        short v[] = {v0, v1, v2, v3, v4, v5, v6, v7};
        *pval = vle16_v_i16m1(v, nlanes);
```
The `pval` points to stack frame, but not to buffer on heap. And it's good luck again, that compiler helps you with UB.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**